src/consensus/blockstore.rs

```diff
@@ -265,70 +138,105 @@ impl Blockstore {
+        if let Some((h, block)) = slot_data.canonical.completed.as_ref() {
+            if *h == hash {
+                return Some(block);
+            }
+        } else {
+            return slot_data
+                .alternatives
+                .get(&hash)
+                .and_then(|data| data.completed.as_ref().map(|(_, block)| Some(block)))
+                .flatten();
+        }
+        return None;
     }
 
     /// Gives reference to stored slice for the given `slot` and `slice` index.
     ///
     /// Returns `None` if blockstore does not hold that slice yet.
     pub fn get_slice(&self, slot: Slot, slice: usize) -> Option<&Slice> {
-        self.slices.get(&(slot, slice))
+        self.slot_data(slot)?.canonical.slices.get(&slice)
     }
 
     /// Gives reference to stored shred for the given `slot`, `slice` and `shred` index.
     ///
     /// Returns `None` if blockstore does not hold that shred yet.
+    // TODO: support alternative/repaired blocks here
     pub fn get_shred(&self, slot: Slot, slice: usize, shred: usize) -> Option<&Shred> {
-        self.shreds
-            .get(&(slot, slice))
+        self.slot_data(slot)?
+            .canonical
+            .shreds
+            .get(&slice)
             .and_then(|v| v.iter().find(|s| s.payload().index_in_slice == shred))
     }
 
-    /// Gives the number of stored slices for a given `slot`.
-    pub fn stored_slices_for_slot(&self, slot: Slot) -> usize {
-        self.slices
-            .range(&(slot, 0)..)
-            .take_while(|(key, _)| key.0 == slot)
-            .count()
-    }
-
-    /// Gives the number of stored shreds for a given `slot` (across all slices).
-    pub fn stored_shreds_for_slot(&self, slot: Slot) -> usize {
-        self.shreds
-            .range(&(slot, 0)..)
-            .take_while(|(key, _)| key.0 == slot)
-            .map(|(_, slice)| slice.len())
-            .sum()
-    }
-
-    /// Gives the number of stored shreds for a given slot and slice.
-    pub fn stored_shreds_for_slice(&self, slot: Slot, slice: usize) -> usize {
-        self.shreds.get(&(slot, slice)).map_or(0, Vec::len)
-    }
-
     /// Generates a Merkle proof for the given `slice` within the given `slot`.
     ///
     /// # Panics
     ///
     /// Panics if the double-Merkle tree for the given `slot` does not exist.
+    // TODO: support alternative/repaired blocks here
     #[must_use]
     pub fn create_double_merkle_proof(&self, slot: Slot, slice: usize) -> Vec<Hash> {
-        let tree = self.double_merkle_trees.get(&slot).unwrap();
+        let slot_data = self.slot_data(slot).unwrap();
+        let tree = slot_data.canonical.double_merkle_tree.as_ref().unwrap();
         tree.create_proof(slice)
     }
+
+    /// Deletes everything before the given `slot` from the blockstore.
+    pub fn prune(&mut self, slot: Slot) {
+        self.block_data = self.block_data.split_off(&slot);
+    }
+
+    ///
+    fn slot_data(&self, slot: Slot) -> Option<&SlotBlockData> {
```

Comment:
I'd be tempted to remove this function as it is not doing a lot and the indirection adds some cognitive load.